File: server/uploadLanggraph.ts
export type UploadProcessingMode =
  | "groupedPaid"
  | "groupedPreview"
  | "singlePaid"
  | "singlePreview";

export function resolveUploadProcessingMode(input: {
  fileCount: number;
  groupAsOne: boolean;
  hasActiveSubscription: boolean;
}): UploadProcessingMode {
  if (input.fileCount < 1) {
    throw new Error("No files uploaded");
  }
  const shouldGroup = input.groupAsOne && input.fileCount > 1;
  if (shouldGroup && input.hasActiveSubscription) return "groupedPaid";
  if (shouldGroup && !input.hasActiveSubscription) return "groupedPreview";
  if (!shouldGroup && input.hasActiveSubscription) return "singlePaid";
  return "singlePreview";
}

File: api-py/app/services/upload_langgraph.py
"""Port of `server/uploadLanggraph.ts`."""

from __future__ import annotations

from typing import Literal

UploadProcessingMode = Literal["groupedPaid", "groupedPreview", "singlePaid", "singlePreview"]


def resolve_upload_processing_mode(
    *,
    file_count: int,
    group_as_one: bool,
    has_active_subscription: bool,
) -> UploadProcessingMode:
    if file_count < 1:
        raise ValueError("No files uploaded")
    should_group = group_as_one and file_count > 1
    if should_group and has_active_subscription:
        return "groupedPaid"
    if should_group and not has_active_subscription:
        return "groupedPreview"
    if not should_group and has_active_subscription:
        return "singlePaid"
    return "singlePreview"
